Plugin fails to load

Currently running mServerLinks 0.0.9-2088c5d on Paper 1.21.1. When the server starts up, this error is thrown:
```
[06:26:14] [ServerMain/ERROR]: [BootstrapProviderStorage] Failed to run bootstrapper for plugins/mServerLinks-Paper-0.0.9-2088c5d.jar. This plugin will not be loaded.
java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:72) ~[?:?]
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502) ~[?:?]
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486) ~[?:?]
	at kotlin.reflect.jvm.internal.calls.CallerImpl$Constructor.call(CallerImpl.kt:41) ~[?:?]
	at kotlin.reflect.jvm.internal.KCallableImpl.callDefaultMethod$kotlin_reflection(KCallableImpl.kt:173) ~[?:?]
	at kotlin.reflect.jvm.internal.KCallableImpl.callBy(KCallableImpl.kt:112) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.kotlin.DataClassFieldDiscoverer$discover$2.complete(ObjectMapping.kt:131) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.kotlin.DataClassFieldDiscoverer$discover$2.complete(ObjectMapping.kt:125) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.objectmapping.ObjectMapperImpl.lambda$load$0(ObjectMapperImpl.java:56) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.objectmapping.ObjectMapperImpl.load0(ObjectMapperImpl.java:115) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.objectmapping.ObjectMapperImpl.load(ObjectMapperImpl.java:56) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.objectmapping.ObjectMapperFactoryImpl.deserialize(ObjectMapperFactoryImpl.java:230) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.serialize.TypeSerializer.deserialize(TypeSerializer.java:96) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.serialize.MapSerializer.deserialize(MapSerializer.java:71) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.serialize.MapSerializer.deserialize(MapSerializer.java:35) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.objectmapping.ObjectMapperImpl.load0(ObjectMapperImpl.java:76) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.objectmapping.ObjectMapperImpl.load(ObjectMapperImpl.java:56) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.objectmapping.ObjectMapperFactoryImpl.deserialize(ObjectMapperFactoryImpl.java:230) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.AbstractConfigurationNode.get0(AbstractConfigurationNode.java:203) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.AbstractConfigurationNode.get(AbstractConfigurationNode.java:161) ~[?:?]
	at dev.mizule.mserverlinks.libs.org.spongepowered.configurate.ConfigurationNode.get(ConfigurationNode.java:521) ~[?:?]
	at mServerLinks-Paper-0.0.9-2088c5d.jar/dev.mizule.mserverlinks.core.config.ConfigurationContainer.load(ConfigurationContainer.java:124) ~[mServerLinks-Paper-0.0.9-2088c5d.jar:?]
	at mServerLinks-Paper-0.0.9-2088c5d.jar/dev.mizule.mserverlinks.paper.mServerLinksBootstrapper.setupConfigs(mServerLinksBootstrapper.java:72) ~[mServerLinks-Paper-0.0.9-2088c5d.jar:?]
	at mServerLinks-Paper-0.0.9-2088c5d.jar/dev.mizule.mserverlinks.paper.mServerLinksBootstrapper.bootstrap(mServerLinksBootstrapper.java:82) ~[mServerLinks-Paper-0.0.9-2088c5d.jar:?]
	at io.papermc.paper.plugin.storage.BootstrapProviderStorage$1.load(BootstrapProviderStorage.java:36) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at io.papermc.paper.plugin.storage.BootstrapProviderStorage$1.load(BootstrapProviderStorage.java:24) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at io.papermc.paper.plugin.entrypoint.strategy.modern.ModernPluginLoadingStrategy.loadProviders(ModernPluginLoadingStrategy.java:117) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at io.papermc.paper.plugin.storage.SimpleProviderStorage.enter(SimpleProviderStorage.java:38) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enter(LaunchEntryPointHandler.java:40) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enterBootstrappers(LaunchEntryPointHandler.java:30) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at net.minecraft.server.Bootstrap.bootStrap(Bootstrap.java:65) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at net.minecraft.server.Main.main(Main.java:124) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at io.papermc.paper.PaperBootstrap.boot(PaperBootstrap.java:21) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at org.bukkit.craftbukkit.Main.main(Main.java:281) ~[paper-1.21.1.jar:1.21.1-52-e08e667]
	at io.papermc.paperclip.Paperclip.lambda$main$0(Paperclip.java:42) ~[app:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.NullPointerException: Parameter specified as non-null is null: method dev.mizule.mserverlinks.core.config.Link.<init>, parameter url
	at mServerLinks-Paper-0.0.9-2088c5d.jar/dev.mizule.mserverlinks.core.config.Link.<init>(Config.kt) ~[mServerLinks-Paper-0.0.9-2088c5d.jar:?]
	at mServerLinks-Paper-0.0.9-2088c5d.jar/dev.mizule.mserverlinks.core.config.Link.<init>(Config.kt:69) ~[mServerLinks-Paper-0.0.9-2088c5d.jar:?]
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62) ~[?:?]
	... 35 more

```

--- a/core/src/main/java/dev/mizule/mserverlinks/core/config/ConfigurationContainer.java
+++ b/core/src/main/java/dev/mizule/mserverlinks/core/config/ConfigurationContainer.java
@@ -30,7 +30,6 @@ import dev.mizule.mserverlinks.core.model.ServerLinkType;
 import io.leangen.geantyref.TypeToken;
 import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
-import org.spongepowered.configurate.CommentedConfigurationNode;
 import org.spongepowered.configurate.ConfigurateException;
 import org.spongepowered.configurate.ConfigurationNode;
 import org.spongepowered.configurate.hocon.HoconConfigurationLoader;
@@ -65,6 +64,7 @@ import java.util.concurrent.atomic.AtomicReference;
  * KickRedirect: https://github.com/4drian3d/KickRedirect/
  */
 public final class ConfigurationContainer<C> {
+
     private final AtomicReference<C> config;
     private final ConfigurationLoader<? extends ConfigurationNode> loader;
     private final Class<C> clazz;
@@ -74,7 +74,7 @@ public final class ConfigurationContainer<C> {
     private ConfigurationContainer(
         final C config,
         final Class<C> clazz,
-        final ConfigurationLoader<? extends ConfigurationNode> loader,
+        final ConfigurationLoader<? extends @NotNull ConfigurationNode> loader,
         final Logger logger
     ) {
         this.config = new AtomicReference<>(config);
@@ -89,7 +89,11 @@ public final class ConfigurationContainer<C> {
             logger.warn("The config file didn't have a version set, assuming latest");
             version.raw(latestVersion);
         } else if (version.getInt() > latestVersion) {
-            logger.error("Loading a newer configuration than is supported ({} > {})! You may have to backup & delete your config file to start the server.", version.getInt(), latestVersion);
+            logger.error(
+                "Loading a newer configuration than is supported ({} > {})! You may have to backup & delete your config file to start the server.",
+                version.getInt(),
+                latestVersion
+            );
         }
     }
 
@@ -129,25 +133,23 @@ public final class ConfigurationContainer<C> {
                     builder.registerAnnotatedObjects(ObjectMappingKt.objectMapperFactory());
                 })
             )
-
             .path(path)
             .build();
 
         try {
             ConfigurationNode node;
             if (Files.notExists(path)) {
-                node = CommentedConfigurationNode.root(loader.defaultOptions());
+                node = loader.load();
                 node.node(ConfigurationContainer.VERSION_FIELD).raw(Transformations.VERSION_LATEST);
             } else {
-                node = loader.load();
+                node = Transformations.updateNode(loader.load());
                 veryConfigVersion(logger, node, Transformations.VERSION_LATEST);
             }
-            node = Transformations.updateNode(node);
             final C config = node.get(clazz);
             node.set(clazz, config);
             loader.save(node);
             return new ConfigurationContainer<>(config, clazz, loader, logger);
-        } catch (ConfigurateException exception){
+        } catch (ConfigurateException exception) {
             logger.error("Could not load {} configuration file", clazz.getSimpleName(), exception);
             return null;
         }
@@ -165,6 +167,7 @@ public final class ConfigurationContainer<C> {
                 .shouldCopyDefaults(true)
                 .header("mServerLinks | by powercas_gamer\n")
                 .serializers(builder -> {
+                    builder.register(TypeToken.get(ServerLinkType.class), new ServerLinkTypeSerializer());
                     builder.registerAnnotatedObjects(org.spongepowered.configurate.kotlin.ObjectMappingKt.objectMapperFactory());
                 })
             )
@@ -174,18 +177,17 @@ public final class ConfigurationContainer<C> {
         try {
             ConfigurationNode node;
             if (Files.notExists(path)) {
-                node = CommentedConfigurationNode.root(loader.defaultOptions());
+                node = loader.load();
                 node.node(ConfigurationContainer.VERSION_FIELD).raw(Transformations.VERSION_LATEST);
             } else {
-                node = loader.load();
+                node = Transformations.updateNode(loader.load());
                 veryConfigVersion(logger, node, Transformations.VERSION_LATEST);
             }
-            Transformations.updateNode(node);
             final C config = node.get(clazz);
             node.set(clazz, config);
             loader.save(node);
             return new ConfigurationContainer<>(config, clazz, loader, logger);
-        } catch (ConfigurateException exception){
+        } catch (ConfigurateException exception) {
             logger.error("Could not load {} configuration file", clazz.getSimpleName(), exception);
             return null;
         }
@@ -225,12 +227,14 @@ public final class ConfigurationContainer<C> {
             .build();
 
         try {
-            final ConfigurationNode node = oldFormat.load();
+            ConfigurationNode node;
+            node = Transformations.updateNode(oldFormat.load());
+            veryConfigVersion(logger, node, Transformations.VERSION_LATEST);
             final C config = node.get(clazz);
             node.set(clazz, config);
             newFormat.save(node);
             return new ConfigurationContainer<>(config, clazz, newFormat, logger);
-        } catch (ConfigurateException exception){
+        } catch (ConfigurateException exception) {
             logger.error("Could not load {} configuration file", clazz.getSimpleName(), exception);
             return null;
         }

--- a/core/src/main/java/dev/mizule/mserverlinks/core/config/ConfigurationContainer.java
+++ b/core/src/main/java/dev/mizule/mserverlinks/core/config/ConfigurationContainer.java
@@ -25,10 +25,12 @@
 package dev.mizule.mserverlinks.core.config;
 
 import dev.mizule.mserverlinks.core.config.serializer.ServerLinkTypeSerializer;
+import dev.mizule.mserverlinks.core.config.transformations.Transformations;
 import dev.mizule.mserverlinks.core.model.ServerLinkType;
 import io.leangen.geantyref.TypeToken;
 import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
+import org.spongepowered.configurate.CommentedConfigurationNode;
 import org.spongepowered.configurate.ConfigurateException;
 import org.spongepowered.configurate.ConfigurationNode;
 import org.spongepowered.configurate.hocon.HoconConfigurationLoader;
@@ -37,6 +39,7 @@ import org.spongepowered.configurate.loader.ConfigurationLoader;
 import org.spongepowered.configurate.yaml.NodeStyle;
 import org.spongepowered.configurate.yaml.YamlConfigurationLoader;
 
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.atomic.AtomicReference;
@@ -66,6 +69,7 @@ public final class ConfigurationContainer<C> {
     private final ConfigurationLoader<? extends ConfigurationNode> loader;
     private final Class<C> clazz;
     private final Logger logger;
+    public static final String VERSION_FIELD = "_version";
 
     private ConfigurationContainer(
         final C config,
@@ -77,6 +81,16 @@ public final class ConfigurationContainer<C> {
         this.loader = loader;
         this.clazz = clazz;
         this.logger = logger;
+    }
+
+    public static void veryConfigVersion(final Logger logger, final ConfigurationNode globalNode, final int latestVersion) {
+        final ConfigurationNode version = globalNode.node(ConfigurationContainer.VERSION_FIELD);
+        if (version.virtual()) {
+            logger.warn("The config file didn't have a version set, assuming latest");
+            version.raw(latestVersion);
+        } else if (version.getInt() > latestVersion) {
+            logger.error("Loading a newer configuration than is supported ({} > {})! You may have to backup & delete your config file to start the server.", version.getInt(), latestVersion);
+        }
     }
 
     public CompletableFuture<Boolean> reload() {
@@ -120,7 +134,15 @@ public final class ConfigurationContainer<C> {
             .build();
 
         try {
-            final ConfigurationNode node = loader.load();
+            ConfigurationNode node;
+            if (Files.notExists(path)) {
+                node = CommentedConfigurationNode.root(loader.defaultOptions());
+                node.node(ConfigurationContainer.VERSION_FIELD).raw(Transformations.VERSION_LATEST);
+            } else {
+                node = loader.load();
+                veryConfigVersion(logger, node, Transformations.VERSION_LATEST);
+            }
+            node = Transformations.updateNode(node);
             final C config = node.get(clazz);
             node.set(clazz, config);
             loader.save(node);
@@ -146,12 +168,19 @@ public final class ConfigurationContainer<C> {
                     builder.registerAnnotatedObjects(org.spongepowered.configurate.kotlin.ObjectMappingKt.objectMapperFactory());
                 })
             )
-
             .path(path)
             .build();
 
         try {
-            final ConfigurationNode node = loader.load();
+            ConfigurationNode node;
+            if (Files.notExists(path)) {
+                node = CommentedConfigurationNode.root(loader.defaultOptions());
+                node.node(ConfigurationContainer.VERSION_FIELD).raw(Transformations.VERSION_LATEST);
+            } else {
+                node = loader.load();
+                veryConfigVersion(logger, node, Transformations.VERSION_LATEST);
+            }
+            Transformations.updateNode(node);
             final C config = node.get(clazz);
             node.set(clazz, config);
             loader.save(node);

--- a/core/src/main/java/dev/mizule/mserverlinks/core/config/transformations/Transformations.java
+++ b/core/src/main/java/dev/mizule/mserverlinks/core/config/transformations/Transformations.java
@@ -25,26 +25,19 @@
 package dev.mizule.mserverlinks.core.config.transformations;
 
 import dev.mizule.mserverlinks.core.config.ConfigurationContainer;
+import org.slf4j.LoggerFactory;
 import org.spongepowered.configurate.ConfigurateException;
 import org.spongepowered.configurate.ConfigurationNode;
-import org.spongepowered.configurate.hocon.HoconConfigurationLoader;
 import org.spongepowered.configurate.serialize.SerializationException;
 import org.spongepowered.configurate.transformation.ConfigurationTransformation;
 
 import java.net.URI;
-import java.nio.file.Paths;
 
 import static org.spongepowered.configurate.NodePath.path;
 
-/**
- * An example of how to use transformations to migrate a configuration to a
- * newer schema version.
- *
- * <p>It's like DFU but not hot garbage! (and probably less PhD-worthy)</p>
- */
 public final class Transformations {
 
-    public static final int VERSION_LATEST = 2; // easy way to track the latest version, update as more revisions are added
+    public static final int VERSION_LATEST = 1;
 
     private Transformations() {}
 
@@ -57,10 +50,9 @@ public final class Transformations {
      */
     public static ConfigurationTransformation.Versioned create() {
         return ConfigurationTransformation.versionedBuilder()
-            .addVersion(VERSION_LATEST, test()) // syntax: target version, latest version
-            .addVersion(1, zeroToOne())
+            .addVersion(VERSION_LATEST, zeroToOne())
             .addVersion(0, initialTransform())
-            .versionKey(path(ConfigurationContainer.VERSION_FIELD))
+            .versionKey(ConfigurationContainer.VERSION_FIELD)
             .build();
     }
 
@@ -72,10 +64,8 @@ public final class Transformations {
      */
     public static ConfigurationTransformation initialTransform() {
         return ConfigurationTransformation.builder()
-            // Move the node at `serverVersion` to the location <code>{"server", "version"}</code>
             .addAction(path(), (path, value) -> {
                 if (value.node(ConfigurationContainer.VERSION_FIELD).virtual()) {
-                    // Add the "config-version" field and set it to 1
                     value.node(ConfigurationContainer.VERSION_FIELD).set(Integer.class, 1);
                 }
                 return null;
@@ -85,7 +75,6 @@ public final class Transformations {
 
     public static ConfigurationTransformation zeroToOne() {
         return ConfigurationTransformation.builder()
-            // oh, turns out we want to use a different format for this, so we'll change it again
             .addAction(path("links"), (path, value) -> {
                 transformUriToUrl(value);
                 return null;
@@ -97,29 +86,13 @@ public final class Transformations {
             .build();
     }
 
-    public static ConfigurationTransformation test() {
-        return ConfigurationTransformation.builder()
-            // oh, turns out we want to use a different format for this, so we'll change it again
-            .addAction(path("update-checker"), (path, value) -> {
-                return null;
-            })
-            .addAction(path("update-checker-links"), (path, value) -> {
-                return null;
-            })
-            .build();
-    }
-
     private static void transformUriToUrl(ConfigurationNode node) {
-        // Iterate over each child node (each Link entry in links or playerLinks map)
         for (ConfigurationNode entry : node.childrenMap().values()) {
-            // Check if "uri" exists and "url" does not
             if (!entry.node("uri").virtual() && entry.node("url").virtual()) {
                 try {
-                    // Get the URI and convert it to a String, set the "url" field
                     String uriString = entry.node("uri").get(URI.class).toString();
                     entry.node("url").set(uriString);
 
-                    // Remove the "uri" field to clean up the config
                     entry.node("uri").set(null);
                 } catch (final SerializationException e) {
                     throw new RuntimeException("Error converting URI to URL in node " + node, e);
@@ -141,37 +114,15 @@ public final class Transformations {
      * @return provided node, after transformation
      */
     public static <N extends ConfigurationNode> N updateNode(final N node) throws ConfigurateException {
-        if (!node.virtual()) { // we only want to migrate existing data
+        if (!node.virtual()) {
             final ConfigurationTransformation.Versioned trans = create();
             final int startVersion = trans.version(node);
-            System.out.println(trans.latestVersion()); // prints 2
-            System.out.println(trans.versionKey().toString()); // prints [[_version]]
             trans.apply(node);
             final int endVersion = trans.version(node);
-            if (startVersion != endVersion) { // we might not have made any changes
-                System.out.println("Updated config schema from " + startVersion + " to " + endVersion);
+            if (startVersion != endVersion) {
+                LoggerFactory.getLogger("mServerLinks").info("Updated config schema from {} to {}", startVersion, endVersion);
             }
         }
         return node;
     }
-
-    public static void main(final String[] args) throws ConfigurateException {
-        if (args.length != 1) {
-            System.err.println("Not enough arguments, usage: transformations <file>");
-            System.err.println("Apply the test transformations to a single file");
-        }
-        final HoconConfigurationLoader loader = HoconConfigurationLoader.builder()
-            .path(Paths.get(args[0]))
-            .build();
-
-        try {
-            loader.save(updateNode(loader.load())); // tada
-        } catch (final ConfigurateException ex) {
-            // We try to update as much as possible, so could theoretically save a partially updated file here
-            System.err.println("Failed to fully update node: " + ex.getMessage());
-            ex.printStackTrace();
-            System.exit(1);
-        }
-    }
-
 }

--- a/core/src/main/java/dev/mizule/mserverlinks/core/config/transformations/Transformations.java
+++ b/core/src/main/java/dev/mizule/mserverlinks/core/config/transformations/Transformations.java
@@ -1,0 +1,177 @@
+/*
+ * This file is part of mServerLinks, licensed under the MIT License.
+ *
+ * Copyright (c) 2024 powercas_gamer
+ * Copyright (c) 2024 contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package dev.mizule.mserverlinks.core.config.transformations;
+
+import dev.mizule.mserverlinks.core.config.ConfigurationContainer;
+import org.spongepowered.configurate.ConfigurateException;
+import org.spongepowered.configurate.ConfigurationNode;
+import org.spongepowered.configurate.hocon.HoconConfigurationLoader;
+import org.spongepowered.configurate.serialize.SerializationException;
+import org.spongepowered.configurate.transformation.ConfigurationTransformation;
+
+import java.net.URI;
+import java.nio.file.Paths;
+
+import static org.spongepowered.configurate.NodePath.path;
+
+/**
+ * An example of how to use transformations to migrate a configuration to a
+ * newer schema version.
+ *
+ * <p>It's like DFU but not hot garbage! (and probably less PhD-worthy)</p>
+ */
+public final class Transformations {
+
+    public static final int VERSION_LATEST = 2; // easy way to track the latest version, update as more revisions are added
+
+    private Transformations() {}
+
+    /**
+     * Create a new builder for versioned configurations. This builder uses a
+     * field in the node (by default {@code schema-version}) to determine the
+     * current schema version (using -1 for no version present).
+     *
+     * @return versioned transformation
+     */
+    public static ConfigurationTransformation.Versioned create() {
+        return ConfigurationTransformation.versionedBuilder()
+            .addVersion(VERSION_LATEST, test()) // syntax: target version, latest version
+            .addVersion(1, zeroToOne())
+            .addVersion(0, initialTransform())
+            .versionKey(path(ConfigurationContainer.VERSION_FIELD))
+            .build();
+    }
+
+    /**
+     * A transformation. This one has multiple actions, and demonstrates how
+     * wildcards work.
+     *
+     * @return created transformation
+     */
+    public static ConfigurationTransformation initialTransform() {
+        return ConfigurationTransformation.builder()
+            // Move the node at `serverVersion` to the location <code>{"server", "version"}</code>
+            .addAction(path(), (path, value) -> {
+                if (value.node(ConfigurationContainer.VERSION_FIELD).virtual()) {
+                    // Add the "config-version" field and set it to 1
+                    value.node(ConfigurationContainer.VERSION_FIELD).set(Integer.class, 1);
+                }
+                return null;
+            })
+            .build();
+    }
+
+    public static ConfigurationTransformation zeroToOne() {
+        return ConfigurationTransformation.builder()
+            // oh, turns out we want to use a different format for this, so we'll change it again
+            .addAction(path("links"), (path, value) -> {
+                transformUriToUrl(value);
+                return null;
+            })
+            .addAction(path("player-links"), (path, value) -> {
+                transformUriToUrl(value);
+                return null;
+            })
+            .build();
+    }
+
+    public static ConfigurationTransformation test() {
+        return ConfigurationTransformation.builder()
+            // oh, turns out we want to use a different format for this, so we'll change it again
+            .addAction(path("update-checker"), (path, value) -> {
+                return null;
+            })
+            .addAction(path("update-checker-links"), (path, value) -> {
+                return null;
+            })
+            .build();
+    }
+
+    private static void transformUriToUrl(ConfigurationNode node) {
+        // Iterate over each child node (each Link entry in links or playerLinks map)
+        for (ConfigurationNode entry : node.childrenMap().values()) {
+            // Check if "uri" exists and "url" does not
+            if (!entry.node("uri").virtual() && entry.node("url").virtual()) {
+                try {
+                    // Get the URI and convert it to a String, set the "url" field
+                    String uriString = entry.node("uri").get(URI.class).toString();
+                    entry.node("url").set(uriString);
+
+                    // Remove the "uri" field to clean up the config
+                    entry.node("uri").set(null);
+                } catch (final SerializationException e) {
+                    throw new RuntimeException("Error converting URI to URL in node " + node, e);
+                } catch (final IllegalArgumentException e) {
+                    throw new RuntimeException("Invalid URI in node " + node, e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Apply the transformations to a node.
+     *
+     * <p>This method also prints information about the version update that
+     * occurred</p>
+     *
+     * @param node the node to transform
+     * @param <N> node type
+     * @return provided node, after transformation
+     */
+    public static <N extends ConfigurationNode> N updateNode(final N node) throws ConfigurateException {
+        if (!node.virtual()) { // we only want to migrate existing data
+            final ConfigurationTransformation.Versioned trans = create();
+            final int startVersion = trans.version(node);
+            System.out.println(trans.latestVersion()); // prints 2
+            System.out.println(trans.versionKey().toString()); // prints [[_version]]
+            trans.apply(node);
+            final int endVersion = trans.version(node);
+            if (startVersion != endVersion) { // we might not have made any changes
+                System.out.println("Updated config schema from " + startVersion + " to " + endVersion);
+            }
+        }
+        return node;
+    }
+
+    public static void main(final String[] args) throws ConfigurateException {
+        if (args.length != 1) {
+            System.err.println("Not enough arguments, usage: transformations <file>");
+            System.err.println("Apply the test transformations to a single file");
+        }
+        final HoconConfigurationLoader loader = HoconConfigurationLoader.builder()
+            .path(Paths.get(args[0]))
+            .build();
+
+        try {
+            loader.save(updateNode(loader.load())); // tada
+        } catch (final ConfigurateException ex) {
+            // We try to update as much as possible, so could theoretically save a partially updated file here
+            System.err.println("Failed to fully update node: " + ex.getMessage());
+            ex.printStackTrace();
+            System.exit(1);
+        }
+    }
+
+}

--- a/core/src/main/kotlin/dev/mizule/mserverlinks/core/config/Config.kt
+++ b/core/src/main/kotlin/dev/mizule/mserverlinks/core/config/Config.kt
@@ -24,6 +24,7 @@
  */
 package dev.mizule.mserverlinks.core.config
 
+import dev.mizule.mserverlinks.core.config.transformations.Transformations
 import dev.mizule.mserverlinks.core.model.ServerLinkType
 import org.spongepowered.configurate.objectmapping.ConfigSerializable
 import org.spongepowered.configurate.objectmapping.meta.Comment
@@ -62,6 +63,9 @@ data class Config(
     val updateChecker: Boolean = true,
     @Setting(value = "bStats")
     val bStats: Boolean = true,
+
+    @Setting(value = ConfigurationContainer.VERSION_FIELD)
+    val version: Int = Transformations.VERSION_LATEST,
 )
 
 @ConfigSerializable

--- a/core/src/main/kotlin/dev/mizule/mserverlinks/core/config/Config.kt
+++ b/core/src/main/kotlin/dev/mizule/mserverlinks/core/config/Config.kt
@@ -65,6 +65,7 @@ data class Config(
     val bStats: Boolean = true,
 
     @Setting(value = ConfigurationContainer.VERSION_FIELD)
+    @Comment("The version of the configuration file\nDo not change this value\nThis is used for internal purposes")
     val version: Int = Transformations.VERSION_LATEST,
 )
 

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -1,6 +1,6 @@
 [versions]
 # minecraft
-mizule = "0.1.14"
+mizule = "0.1.15"
 minecraft = "1.21"
 hangar-publish = "0.1.2"
 desertwell = "2.0.4"

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.10-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.10.1-bin.zip
 networkTimeout=10000
 validateDistributionUrl=true
 zipStoreBase=GRADLE_USER_HOME

